Assets/Scripts/Achievement/AchievementManager.cs

```diff
@@ -13,73 +15,114 @@ public class AchievementManager : Singleton<AchievementManager>
+                }
             }
+        }
 
+        private void InvokeDefaultEvents()
+        {
+            UIManager.Instance.ScoreIncreased += ScoreIncreased;
+            Enemy.EnemyManager.Instance.EnemyDestroyed += EnemyKilled;
+            BTManager.GameManager.Instance.GameStarted += GamesPlayed;
         }
 
         private void ScoreIncreased()
         {
-            hiScoreController.CheckAchievement(UIManager.Instance.playerScore);
+            CheckForAchievement(AchievementType.hiScore, UIManager.Instance.playerScore);
         }
 
         private void EnemyKilled()
         {
-            enemiesKilledController.CheckAchievement(Enemy.EnemyManager.Instance.enemiesKilled);
+            CheckForAchievement(AchievementType.enemyKilled, Enemy.EnemyManager.Instance.enemiesKilled);
         }
 
         private void GamesPlayed()
         {
-            gamesPlayedController.CheckAchievement(BTManager.GameManager.Instance.gamesPlayed);
+            CheckForAchievement(AchievementType.gamesPlayed, BTManager.GameManager.Instance.gamesPlayed);
         }
 
-        public void CheckAchievement(AchievementType achievementType, int achievementIndex)
+        void CheckForAchievement(AchievementType achievementType, int achievedVal)
         {
-            AchievementCheck?.Invoke(achievementType, achievementIndex);
-        }
-
-        private void CallAchievementUI(AchievementType achievementType, int achievementIndex)
-        {
-            for (int i = 0; i < achievementScriptable.achievements.Count; i++)
+            for (int i = 0; i < achievementList.Count; i++)
             {
-                if(achievementScriptable.achievements[i].achievementType == achievementType)
+                if (CheckAchievementType(achievementType, i))
                 {
-                    for (int k = 0; k < achievementScriptable.achievements[i].achievementInfo.Count; k++)
+                    if (AchievementLock(i) && AchievementThreshHold(achievedVal, i))
                     {
-                        if(k == achievementIndex)
-                        {
-                            string value = "Unlocked " + achievementType.ToString() + " Achievement. Title " +
-                                          achievementScriptable.achievements[i].achievementInfo[k].achievementTitle;
-                            AchievementUI?.Invoke(value);
-                        }
+                        string value = "Unlocked " + achievementType.ToString() + " Achievement. Title " +
+                                              achievementList[i].achievementInfo.achievementTitle;
+
+                        Achievement achievement = new Achievement();
+                        achievement = achievementList[i];
+                        achievement.achievementInfo.achievementStatus = AchievementStatus.Unlocked;
+
+                        achievementList[i] = achievement;
+                        AchievementUnlocked?.Invoke(value);
```

Comment:
why do we have 2 events fired at the same place?

Reply:
One is to display achievement in game scene when unlocked
One is to check reward for specific player